client/admin/src/components/Table.vue

```diff
@@ -0,0 +1,353 @@
+                  v-for="(val, key) in values"
+                  :key="key"
+                  style="white-space: pre"
+                >
+                  {{ key }} : {{ val }}
+                </div>
+              </div>
+            </td>
+          </tr>
+        </tbody>
+      </table>
+    </div>
+
+    <div class="inline-flex justify-center content-center items-center">
+      <button
+        id="pprev"
+        @click="paginatorPrevious"
+        class="bg-gray-300 hover:bg-gray-400 text-gray-800 font-bold py-2 px-4 rounded-l"
+      >
+        Prev
+      </button>
+
+      <button
+        id="pnext"
+        @click="paginatorNext"
+        class="bg-gray-300 hover:bg-gray-400 text-gray-800 font-bold py-2 px-4 rounded-r"
+      >
+        Next
+      </button>
+      <input
+        class="shadow appearance-none border rounded py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline w-16"
+        id="ppage"
+        type="text"
+        placeholder="Page"
+        v-model="paginatorInfo.currentPage"
+        @change="paginatorInputChange"
+      />
+      <label for="page"> of {{ paginatorInfo.totalPages }}</label>
+    </div>
+
+    <button
+      @click="showSettings = !showSettings"
+      class="bg-white hover:bg-gray-100 text-gray-800 font-semibold py-2 px-4 border border-gray-400 rounded shadow my-4 mx-6"
+    >
+      <p v-if="!showSettings">Show settings</p>
+      <p v-if="showSettings">Close settings</p>
+    </button>
+
+    <transition
+      enter-active-class="transition duration-1000 ease-out"
+      leave-active-class="transition duration-75 ease-in"
+      enter-class="transform opacity-0 scale-95"
+      enter-to-class="transform opacity-100 scale-100"
+      leave-class="transform opacity-100 scale-100"
+      leave-to-class="transform opacity-0 scale-95"
+    >
+      <div v-show="showSettings" class="bg-gray-500 rounded p-6 flex flex-wrap">
+        <div v-for="(option, key) in productAttributes" :key="key" class="p-4">
+          <input
+            type="checkbox"
+            :id="option"
+            :name="option"
+            :value="option"
+            @click="checked"
+            :checked="key < 4"
+          />
+          <label :for="option" class="p-2">{{ option }}</label
+          ><br />
+        </div>
+      </div>
+    </transition>
+  </div>
+</template>
+
+<script lang="ts">
+import { Component, Vue } from "vue-property-decorator";
+import ErrorNotification from "@/components/ErrorNotification.vue";
+import SuccessNotification from "@/components/SuccessNotification.vue";
+import { Action, Getter, namespace } from "vuex-class";
+
+const TableAction = namespace("table", Action);
+const TableGetter = namespace("table", Getter);
+
+@Component({
+  components: {
+    ErrorNotification,
+    SuccessNotification,
+  },
+})
+export default class Table extends Vue {
+  private errors: any[] = [];
+  private showSettings: Boolean = false;
+
+  private products: any[] = [];
```

Comment:
@AlexeiaW would it be possible to make the table more generic in terms of defining what data is contained, bearing in mind we would want to support various data types like users, orders, etc. within tables in future, so the data is really more defined like `rows` and `columns`. Maybe a prop could handle data transformation rather than transformation being built into the table itself, kind of like a contract. Let me know what you think.

In this sense, there would be no reference to products inside the table. It would reference a generic variable like `dataRow` or something like that, which could be a row of product data or a row of user data, or any other tabular based data

Reply:
@luke-siedle yes sure, I will give that a go and try to make the table more generic as you described 👍